add local to prod example (#8948)
dagster-io/dagster: docs/content/guides/dagster/local_to_production_tutorial.mdx

---
title: Development to Production Workflow | Dagster
description: This guide walks through how to transition your code from local development to production
---

# Development to Production Workflow

> In this example, we will walk through how to transition your assets and jobs from local development, to unit testing, and production deployments. We will cover swapping out resources and IO managers and setting run config based on the environment where your code will be executed.

Reviewer: ```suggestion
In this example, we will walk through how to transition your data pipelines from local development, to unit testing, and production deployments. We will cover swapping out resources and IO managers and setting run config based on the environment where your code will be executed.
```
I feel this guide may get referenced by someone when they're talking about general data engineering best practices (in other words, the audience of this page might be very general data practitioners who are not familiar with dagster), so "assets and jobs" may not be the most familiar term for general audience.

> In this example, we will walk through how to transition your assets and jobs from local development, to unit testing, and production deployments. We will cover swapping out resources and IO managers and setting run config based on the environment where your code will be executed.

Author: the line doesn't exist anymore, resolving

If you’re new to dagster, here’s an overview of the main concepts we’ll be using in this example (if you’re already familiar with dagster, feel free to skip to the next section):

- asset - an asset is a software object that models a data asset. The prototypical example is a table in a database or a file in cloud storage.
- resource - a resource is an object that models a connection to a (typically) external service. Resources can be shared between assets and different implementations of resources can be swapped in depending on environment. For example, a resource may provide methods to send messages in Slack.
- IO manager - an IO manager is a special kind of resource that handles storing and loading assets. For example, if we wanted to store assets in S3, we would use dagster’s built in S3 IO manager.

> - run config - assets and resources sometimes require configuration to set certain values (like the password to a database). Run config allows you to set these values at run time. In this example, we will also use an API to set some default run configuration.

Reviewer: let's make sure we link to the relevant concept pages for further navigation.

## Problem Statement

Let’s say we’ve been tasked with fetching data from Hacker News and splitting it into two datasets: one containing all of the data about stories and one containing all of the data about comments. We have the following additional requirements:

- For local development, we want to store the datasets as tables in a development snowflake database called `SANDBOX` that has a schema for each user.
- For production, we want to store the datasets as tables in a production snowflake database called `PRODUCTION` that has a schema called `HACKER_NEWS`.

> - For testing, we don’t want to fetch data from Hacker News, instead we want to provide seed data so we can assert that the data was split into the stories and comments datasets correctly.

Reviewer: I feel we could make this section above the "If you're new to dagster...", and maybe center it around the "why's" such as the current pain points that you can't simply run the same code across different environments without the fear of breaking something that you are not aware of. 

i think the sda guide ([why use sda](https://dagster-btbtnhnr6-elementl.vercel.app/guides/dagster/enriching-with-software-defined-assets#why-use-software-defined-assets)) did a good job attracting audience who has lived in the same pain. we don't have to expand too much like the sda guide did but several painful scenarios might help the reader understand why a good workflow matters to them.

cc @erinkcochran87 may have thoughts on how we structuring the info too.

Reviewer 2: I agree with bumping this section up and including pain points. Perhaps the general structure could be something like this:

- Pain points, such as not being able to run the same code across environments
- Dagster concepts used in the guide and how they help address pain points
- Brief summary of the example we'll use to demonstrate the pain points and how to solve them

Author: i reorganized this section based on your ideas, let me know what you think!

> ### Setup

Reviewer 2: I feel like we should call out the libraries/modules being used in this section. I got to the callout about Pandas dataframes before I realized that was in the guide.

I'm not sure if other folks will care, but I prefer it when guides like this spell everything out for me.

Author: I added a reference to pandas before the first asset code snippet. does that read better to you? I could also add it to the section of bullet points where i say 
```
Using these Dagster concepts we will:
* write three assets: one to represent the Hacker News data and two for the datasets containing data for comments and stories
* write a resource to fetch data from Hacker News
* use Dagster's Snowflake IO manager to store the datasets in [Snowflake](https://www.snowflake.com/)
    - for local development, we will use a Snowflake database called `SANDBOX` with a schema for each user
    - for staging, we will use a Snowflake database called `STAGING` with a `HACKER_NEWS` schema
    - for production, we will use a Snowflake database called `PRODUCTION` with a `HACKER_NEWS` schema
* set up our repository so that the configuration for the Snowflake IO manager is automatically supplied based on the environment where the code is running
```

If you want to follow along with this example, you can either:

Clone the example (this will give you the final version of the code):

    dagster new-project --example local_to_prod
    cd local_to_prod
    pip install .

Start with a scaffolded project and fill in the code as we go

>     dagster new-project --quick-start local_to_prod

Reviewer 2: By having this command come first, does it assume the user already has Dagster installed?

Sorry if it's obvious - still pretty new to this! I'm confused because `dagster` is here but the last line has `pip install dagster ...`

Author: good catch, that's an error on my part. I have a TODO to run both of these setups before the PR lands

    cd local_to_prod
    pip install dagster dagit pandas pytest dagster-snowflake dagster-snowflake-pandas

> ### Part One: Local Development

Reviewer 2: ```suggestion
### Part one: Local development
```

Headings should use sentence case :)

> Let’s start by writing the assets

Reviewer 2: This is the summary from the **Problem statement** that describes this section:

> For local development, we want to store the datasets as tables in a development snowflake database called `SANDBOX` that has a schema for each user.

I think it'd be helpful to replace the current sentence (`Let's start by writing the assets`) with something like this that's more specific. It can help add context to the example code that follows.

Author: I liked your suggestion of having a list of what we'll do in each section. I added that to the start of this section to hopefully add the missing context

> 

Reviewer 2: It may also be helpful to list out the specific steps we're going to take in this section - ex:

```
In this section, we'll:

- Write our assets
- Write a resource to fetch data from the Hacker News API
- Add the assets to our repository
- Add run configuration for the I/O manager
- Materialize assets in Dagit
```

There's a lot happening in this section. Providing an outline of what's in it can help ensure readers don't miss anything.

I'm on the fence about whether headings should be added for each step (not sure how I feel about that for overall flow), but it could:

- Make it easier to skim
- Allow users to directly link to steps/examples they care about
- Help users see how far along they are in the section

    """Items from the Hacker News API: each is a story or a comment on a story."""

    hn_client = context.resources.hn_client

    rows = []
    for item_id in range(1, 100 + 1):
        rows.append(hn_client.fetch_item_by_id(item_id))

    result = pd.DataFrame(rows, columns=hn_client.item_field_names).drop_duplicates(subset=["id"])
    result.rename(columns={"by": "user_id"}, inplace=True)

    return result

@asset
def comments(items: pd.DataFrame) -> pd.DataFrame:
    """Comments from the Hacker News API."""
    return items[items["type"] == "comment"]

@asset
def stories(items: pd.DataFrame) -> pd.DataFrame:
    """Stories from the Hacker News API."""
    return items[items["type"] == "story"]
```

We’ll also need to write the resource to fetch data from Hacker News

<Note>

When you’re in the beginning development stages, you may find it easier to write the code to fetch data directly in your assets or ops. This is great for initial development, but eventually you may want to refactor your code to use resources to support things like testing. In this case we’re going to jump right into defining a resource because we know we’ll want to write a mock resource shortly.

> </Note>

Reviewer: nice

```python file=/guides/dagster/local_to_prod_tutorial/repository/repository_v1.py startafter=start endbefore=end
# repository.py

@repository
def repo():
    return [
        *with_resources(
            [items, comments, stories],
            resource_defs={"hn_client": hn_api_client, "io_manager": snowflake_io_manager}
        )
    ]
```

When we materialize the assets in Dagit we can add configuration for the `snowflake_io_manager` in the launchpad or we can use the `configured` API and update our repository to look like this:

```python file=/guides/dagster/local_to_prod_tutorial/repository/repository_v2.py startafter=start endbefore=end
# repository.py

@repository
def repo():
    resource_defs = {
        "hn_client": hn_api_client,
        "io_manager": snowflake_io_manager.configured(
            {
                "account": "abc1234.us-east-1",
                "user": "me@company.com",
                "password": "my_super_secret_password",
                "database": "SANDBOX",
                "schema": "JAMIE"
            }
        ),
    }

    return [*with_resources([items, comments, stories], resource_defs=resource_defs)]
```

> Now we can materialize the assets the Dagit and ensure that the data appears in snowflake as we expect:

Reviewer 2: ```suggestion
Now we can materialize the assets the Dagit and ensure that the data appears in Snowflake as we expect:
```

<!-- \[Image: Screen Shot 2022-07-19 at 3.21.32 PM.png] -->

<Note>

> While we define our assets as pandas DataFrames, the snowflake IO manager automatically translates the asset data to and from snowflake tables. The Python asset name determines the snowflake table name. In this case three tables will be created: ITEMS, COMMENTS, and STORIES

Reviewer 2: ```suggestion
While we define our assets as Pandas DataFrames, the Snowflake I/O manager automatically translates the asset data to and from Snowflake tables. The Python asset name determines the Snowflake table name. In this case, three tables will be created: <code>ITEMS</code>, <code>COMMENTS</code>, and <code>STORIES</code>
```

- Capitalizing proper nouns
- Adding code formatting to table names

</Note>

### Part Two: Testing

> Let’s write some tests! We don’t want to rely on external calls to the Hacker News API during out unit tests, so the first thing we will need to do is write a version of the Hacker News API client that will serve consistent data.

Reviewer 2: ```suggestion
Let’s write some tests! We don’t want to rely on external calls to the Hacker News API during our unit tests, so the first thing we'll do is write a version of the Hacker News API client that will serve consistent data.
```

<Note>

> At Dagster we believe one of the main barriers to testing data applications is that business logic depends on external, heavy-weight dependencies that are difficult to run in a testing scenario. This is why we developed the concept of [resources](/concepts/resources).

Author: this note got removed in the restructure anyway!

> At Dagster we believe one of the main barriers to testing data applications is that business logic depends on external, heavy-weight dependencies that are difficult to run in a testing scenario. This is why we developed the concept of [resources](/concepts/resources).
> 
> When testing an asset or op with an external dependency (either a resource or IO manager), we recommend using a lightweight version of the resource that implements all of the methods used in the asset or op. You can use a Dagster-provided resource (like the in memory IO manager), or write your own implementation.
> 
> For more information and recommendations on testing, see our [testing documentation](/concepts/testing).
> 

Reviewer 2: ```suggestion
At Dagster we believe one of the main barriers to testing data applications is that business logic depends on external, heavy-weight dependencies that are difficult to run in a testing scenario. This is why we developed the concept of [resources](/concepts/resources).

When testing an asset or op with an external dependency - like a resource or I/O manager - we recommend using a lightweight version of the resource that implements all of the methods used in the asset or op. You can use a Dagster-provided resource (like the in-memory I/O manager) or write your own implementation.

For more information and recommendations on testing, refer to the [testing documentation](/concepts/testing).

```

Suggesting some small tweaks to vary the sentence structure a bit.

Author: this note doesn't exist anymore, resolving

</Note>

> We want to ensure that the version of the Hacker News Client we write for testing follows the API of the real Hacker News Client, so at this point, it would be a good idea to write an interface for the testing resource and `HNAPIClient` to implement:

Reviewer 2: This sentence is a bit long - can we break it up? Ex:

```suggestion
We want to ensure that the testing version of our Hacker News Client follows the API of the real client. To do this, we can write an interface for the testing resource and `HNAPIClient` to implement:
```

Additionally, the `to implement` at the end of this sentence confuses me a bit - what is being implemented? The interface? Something else?

```python file=/guides/dagster/local_to_prod_tutorial/resources/resources_v2.py startafter=start_abc endbefore=end_abc
# resources.py

class HNClient(ABC):
    """
    Base class for a Hacker News Client
    """

    @property
    def item_field_names(self):
        pass

    def fetch_item_by_id(self, item_id: int) -> Optional[Dict[str, Any]]:
        pass

    def fetch_max_item_id(self) -> int:
        pass
```

> then we can write the version of the resource we will use for testing:

Reviewer 2: ```suggestion
Next, we'll write the version of the resource we'll use for testing:
```

Capitalization and editing for conciseness.

        return len(self.data.items())

@resource
def mock_hn_client():
    return MockHNClient()
```

Now, let’s write tests to verify that the assets produce the datasets we expect:

```python file=/guides/dagster/local_to_prod_tutorial/test_assets.py startafter=start endbefore=end
# test_assets.py

def test_items():
    context = build_op_context(resources={"hn_client": mock_hn_client})
    hn_dataset = items(context)
    assert isinstance(hn_dataset, pd.DataFrame)

    expected_data = pd.DataFrame(MockHNClient().data.values()).rename(columns={"by": "user_id"})

    assert (hn_dataset == expected_data).all().all()

def test_comments():
    mock_data = pd.DataFrame({"id": [1, 2, 3], "type": ["comment", "story", "comment"]})
    out_data = comments(mock_data)

    out_ids = out_data["id"].tolist()

    assert 1 in out_ids
    assert 2 not in out_ids
    assert 3 in out_ids

def test_stories():
    mock_data = pd.DataFrame({"id": [1, 2, 3], "type": ["story", "story", "comment"]})
    out_data = stories(mock_data)

    out_ids = out_data["id"].tolist()

    assert 1 in out_ids
    assert 1 in out_ids
    assert 3 not in out_ids
```

### Part Three: Production

> Now that our assets work locally and can be tested, let’s set up for the production environment. We want to store the assets in a production Snowflake database, so we need to update the configuration for the `snowflake_io_manager`:

Reviewer 2: > Now that our assets work locally and can be tested, let’s set up for the production environment.

I'm finding the second phrase in this sentence a bit awkward. Maybe something like `... and can be tested, let's modify our code for a production environment.`

Open to suggestions!

```python file=/guides/dagster/local_to_prod_tutorial/repository/repository_v3.py startafter=start endbefore=end
# repository.py

@repository
def repo():
    resource_defs = {
        "hn_client": hn_api_client,
        "io_manager": snowflake_io_manager.configured(
            {
                "account": "abc1234.us-east-1",
                "user": "dev@company.com",
                "password": "company_super_secret_password",
                "database": "PRODUCTION",
                "schema": "HACKER_NEWS"
            }
        ),
    }

    return [*with_resources([items, comments, stories], resource_defs=resource_defs)]
```

> Here we run into an issue. The next time a developer wants to work on the assets, they will have to remember to change the configuration for the `snowflake_io_manager`. This leaves room for a developer to accidentally override the production asset during local development. Instead we can set up the repository to attach resources to assets based on the environment:

Reviewer 2: ```suggestion
Here we run into an issue. The next time a developer wants to work on the assets, they will have to remember to change the configuration for the `snowflake_io_manager`. This leaves room for a developer to accidentally overwrite the production asset during local development.

Instead, we can set up the repository to attach resources to assets based on the environment:
```

Suggesting `overwrite` because it made more sense to me, but feel free to ignore if it's actually incorrect.

Author: overwrite is correct! thanks for catching it

```python file=/guides/dagster/local_to_prod_tutorial/repository/repository_v4.py startafter=start endbefore=end
# repository.py

@repository
def repo():
    resource_defs = {
        "local": {
            "hn_client": hn_api_client,
            "io_manager": snowflake_io_manager.configured({
                    "account": "abc1234.us-east-1",
                    "user": "me@company.com",
                    "password": "my_super_secret_password",
                    "database": "SANDBOX",
                    "schema": "JAMIE"
            })
        },
        "production": {
            "hn_client": hn_api_client,
            "io_manager": snowflake_io_manager.configured({
                    "account": "abc1234.us-east-1",
                    "user": "dev@company.com",
                    "password": "company_super_secret_password",
                    "database": "PRODUCTION",
                    "schema": "HACKER_NEWS"
            })
        }
    }
    deployment_name = os.environ.get("DAGSTER_DEPLOYMENT", "local")

    return [
        with_resources([items, comments, stories], resource_defs=resource_defs[deployment_name])
    ]
```

> Now we just set the environment variable `DAGSTER_DEPLOYMENT=production` in our deployment and the correct resources will be applied to the assets.

Reviewer 2: ```suggestion
Now, we can set the environment variable `DAGSTER_DEPLOYMENT=production` in our deployment and the correct resources will be applied to the assets.
```

Nit - removing `just` as it's not necessary.

<Note>

If you are using \[Dagster Cloud]\(link to dagster cloud page explaining this?), this environment variable will be set for you. Additionally, if you are using \[Branch Deployments]\(link to branch deployments page explaining this?) `DAGSTER_DEPLOYMENT` will be automatically set to `VAR`.

</Note>

> We still have some problems with this setup:

Reviewer: maybe a h4 here for parametrizing I/O based on envs

Author: i think this comment may have referred to a different line than the one shown in the preview, can you elaborate on what you were meaning?

Reviewer: i forget 😅 i dont see it being an issue anymore

> 1. Each developer still needs to remember to change `user` and `password` to their credentials and `schema` to their name when they develop locally.

Reviewer 2: ```suggestion
1. Developers need to remember to change `user` and `password` to their credentials and `schema` to their name when developing locally.
```

Editing for conciseness/clarity.

2. Passwords are being stored in code.
3. If we want to change the database where we store development or production data, we need to make a code change.

> We can easily solve all of these problems because the snowflake IO manager accepts configuration from environment variables using the <PyObject object="StringSource" /> configuration type. This allows us to store configuration values as environment variables, and point dagster to those environment variables in the run configuration:

Author: do you think linking to api docs for StringSource is too in the weeds?

Reviewer: I think it's good to mention because it is the recommended way to deal with env var.

> We can easily solve all of these problems because the snowflake IO manager accepts configuration from environment variables using the <PyObject object="StringSource" /> configuration type. This allows us to store configuration values as environment variables, and point dagster to those environment variables in the run configuration:

Reviewer 2: ```suggestion
We can easily solve all of these problems because the Snowflake I/O manager accepts configuration from environment variables using the <PyObject object="StringSource" /> configuration type. This allows us to store configuration values as environment variables, and point Dagster to those environment variables in the run configuration:
```

```python file=/guides/dagster/local_to_prod_tutorial/repository/repository_v5.py startafter=start endbefore=end
# repository.py

@repository
def repo():
    resource_defs = {
        "hn_client": hn_api_client,
        "io_manager": snowflake_io_manager.configured(
            {
                "account": {"env": "SNOWFLAKE_ACCOUNT"},
                "user": {"env": "SNOWFLAKE_USER"},
                "password": {"env": "SNOWFLAKE_PASSWORD"},
                "database": {"env": "SNOWFLAKE_DATABASE"},
                "schema": {"env": "SNOWFLAKE_SCHEMA"},
            }
        ),
    }

    return [*with_resources([items, comments, stories], resource_defs=resource_defs)]
```

<Note>

> Since all of the snowflake IO manager configuration is being set through environment variables, we no longer need to select the resource configuration based on deployment! If we later add more configuration that varies between local and production, we can go back to the previous pattern of selecting config based on the deployment name.

Reviewer 2: ```suggestion
Since all of the Snowflake I/O manager configuration is being set through environment variables, we no longer need to select the resource configuration based on deployment! If we decide to add more configuration that varies between local and production, we can go back to the previous pattern of selecting config based on the deployment name.
```
